File: 0907-sum-of-subarray-minimums/0907-sum-of-subarray-minimums.java
class Solution {
    public int sumSubarrayMins(int[] arr) {
        int n = arr.length;
        // Next smallest Element
        int NSE[] = new int[n];
        // Previous Smallest Or Equal
        int PSOE[] = new int[n];
        Stack<Integer> stack = new Stack<>();
        stack.push(0);
        for (int i = 1; i < n; i++) {
            while (!stack.isEmpty() && arr[i] < arr[stack.peek()]) {
                NSE[stack.pop()] = i;
            }
            stack.push(i);
        }
        while (!stack.isEmpty()) {
            NSE[stack.pop()] = n;
        }
        // Now find PSOE
        stack.push(n - 1);
        for (int i = n - 2; i >= 0; i--) {
            while (!stack.isEmpty() && arr[i] <= arr[stack.peek()]) {
                PSOE[stack.pop()] = i;
            }
            stack.push(i);
        }
        while (!stack.isEmpty()) {
            PSOE[stack.pop()] = -1;
        }
        long sum = 0;
        int MOD = 1000000007;
        for (int i = 0; i < n; i++) {
            long a = (i - PSOE[i]);
            long b = (NSE[i] - i);
            long c = arr[i];
            long val = (a*b)%MOD;
            val = (val*c)%MOD;
            sum = (sum+val)%MOD;
        }
        return (int) sum;
    }
}
 

File: 0907-sum-of-subarray-minimums/0907-sum-of-subarray-minimums.py
from collections import deque
class Solution:
    def sumSubarrayMins(self, arr: List[int]) -> int:
        m=10**9 +7
        n=len(arr)
        nse=[0]*n
        psoe=[0]*n
        stk=deque()
        stk.append(0)
        for i in range(1,n):
            while(len(stk)!=0 and arr[i]<arr[stk[-1]]):
                nse[stk[-1]]=i
                stk.pop()
            stk.append(i)
        while(len(stk)!=0):
            nse[stk[-1]]=n
            stk.pop()

        stk.append(n-1)
        for i in range(n-2,-1,-1):
            while(len(stk)!=0 and arr[i]<=arr[stk[-1]]):
                psoe[stk[-1]]=i
                stk.pop()
            stk.append(i)
        while(len(stk)!=0):
            psoe[stk[-1]]=-1
            stk.pop()

        sum1=0
        for i in range(n):
            x1=(i-psoe[i])
            x2=nse[i]-i
            x= (((x1%m * x2%m)%m) * (arr[i]%m))%m
            sum1=(sum1%m + x%m)%m
        return sum1